File: src/main/java/net/kingdommc/darkages/numinoustreasury/command/node/NodeCommand.java
package net.kingdommc.darkages.numinoustreasury.command.node;

import net.kingdommc.darkages.numinoustreasury.NuminousTreasury;
import net.kingdommc.darkages.numinoustreasury.command.node.session.NodeSessionCommand;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static net.md_5.bungee.api.ChatColor.RED;

public final class NodeCommand implements CommandExecutor, TabCompleter {

    private final NodeCreateCommand nodeCreateCommand;
    private final NodeDeleteCommand nodeDeleteCommand;
    private final NodeSessionCommand nodeSessionCommand;

    private final List<String> createAliases = List.of("create", "new", "c", "n");
    private final List<String> deleteAliases = List.of("delete", "remove", "del", "rm", "d", "r");
    private final List<String> sessionAliases = List.of("session", "s");
    private final List<String> subcommands = new ArrayList<>(){{
        addAll(createAliases);
        addAll(deleteAliases);
        addAll(sessionAliases);
    }};

    public NodeCommand(NuminousTreasury plugin) {
        nodeCreateCommand = new NodeCreateCommand(plugin);
        nodeDeleteCommand = new NodeDeleteCommand(plugin);
        nodeSessionCommand = new NodeSessionCommand(plugin);
    }

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if (args.length < 1) {
            sender.sendMessage(RED + "Usage: /node [create|delete|session]");
            return true;
        }
        if (createAliases.contains(args[0].toLowerCase())) {
            return nodeCreateCommand.onCommand(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
        } else if (deleteAliases.contains(args[0].toLowerCase())) {
            return nodeDeleteCommand.onCommand(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
        } else if (sessionAliases.contains(args[0].toLowerCase())) {
            return nodeSessionCommand.onCommand(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
        } else {
            sender.sendMessage(RED + "Usage: /node [create|delete|session]");
            return true;
        }
    }

    @Override
    public List<String> onTabComplete(CommandSender sender, Command command, String label, String[] args) {
        if (args.length < 1) {
            return subcommands;
        } else if (args.length == 1) {
            return subcommands.stream().filter(subcommand -> subcommand.startsWith(args[0].toLowerCase())).toList();
        } else {
            if (createAliases.contains(args[0].toLowerCase())) {
                return nodeCreateCommand.onTabComplete(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
            } else if (deleteAliases.contains(args[0].toLowerCase())) {
                return nodeDeleteCommand.onTabComplete(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
            } else if (sessionAliases.contains(args[0].toLowerCase())) {
                return nodeSessionCommand.onTabComplete(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
            } else {
                return List.of();
            }
        }
    }
}

File: src/main/kotlin/net/arvandor/numinoustreasury/command/node/NodeCommand.kt
package net.arvandor.numinoustreasury.command.node

import net.arvandor.numinoustreasury.NuminousTreasury
import net.arvandor.numinoustreasury.command.node.session.NodeSessionCommand
import net.md_5.bungee.api.ChatColor
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter

class NodeCommand(plugin: NuminousTreasury) : CommandExecutor, TabCompleter {
    private val nodeCreateCommand = NodeCreateCommand(plugin)
    private val nodeDeleteCommand = NodeDeleteCommand(plugin)
    private val nodeSessionCommand = NodeSessionCommand(plugin)

    private val createAliases = listOf("create", "new", "c", "n")
    private val deleteAliases = listOf("delete", "remove", "del", "rm", "d", "r")
    private val sessionAliases = listOf("session", "s")
    private val subcommands = createAliases + deleteAliases + sessionAliases

    override fun onCommand(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>,
    ): Boolean {
        if (args.size < 1) {
            sender.sendMessage(ChatColor.RED.toString() + "Usage: /node [create|delete|session]")
            return true
        }
        if (createAliases.contains(args[0].lowercase())) {
            return nodeCreateCommand.onCommand(
                sender,
                command,
                label,
                args.drop(1).toTypedArray(),
            )
        } else if (deleteAliases.contains(args[0].lowercase())) {
            return nodeDeleteCommand.onCommand(
                sender,
                command,
                label,
                args.drop(1).toTypedArray(),
            )
        } else if (sessionAliases.contains(args[0].lowercase())) {
            return nodeSessionCommand.onCommand(
                sender,
                command,
                label,
                args.drop(1).toTypedArray(),
            )
        } else {
            sender.sendMessage(ChatColor.RED.toString() + "Usage: /node [create|delete|session]")
            return true
        }
    }

    override fun onTabComplete(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>,
    ): List<String>? {
        return if (args.size < 1) {
            subcommands
        } else if (args.size == 1) {
            subcommands
                .filter { subcommand: String -> subcommand.startsWith(args[0].lowercase()) }
        } else {
            if (createAliases.contains(args[0].lowercase())) {
                nodeCreateCommand.onTabComplete(
                    sender,
                    command,
                    label,
                    args.drop(1).toTypedArray(),
                )
            } else if (deleteAliases.contains(args[0].lowercase())) {
                nodeDeleteCommand.onTabComplete(
                    sender,
                    command,
                    label,
                    args.drop(1).toTypedArray(),
                )
            } else if (sessionAliases.contains(args[0].lowercase())) {
                nodeSessionCommand.onTabComplete(
                    sender,
                    command,
                    label,
                    args.drop(1).toTypedArray(),
                )
            } else {
                listOf<String>()
            }
        }
    }
}
